Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/calibration/ErrorHandler.java

```diff
@@ -0,0 +1,519 @@
+            rmseLowDeviations[i] = new Deviation(1.0 / errorHorizon);
+            intervalPrecision[i] = new Deviation(1.0 / errorHorizon);
+        }
+        errorMean = new float[length];
+        errorRMSE = new DiVector(length);
+        lastDataDeviations = new float[inputLength];
+        errorDistribution = new RangeVector(length);
+        if (builder.upperLimit.isPresent()) {
+            checkArgument(builder.upperLimit.get().length == inputLength, "incorrect length");
+            upperLimit = Arrays.copyOf(builder.upperLimit.get(), inputLength);
+        } else {
+            upperLimit = new float[inputLength];
+            Arrays.fill(upperLimit, Float.MAX_VALUE);
+        }
+        if (builder.lowerLimit.isPresent()) {
+            checkArgument(builder.lowerLimit.get().length == inputLength, "incorrect length");
+            for (int y = 0; y < inputLength; y++) {
+                checkArgument(builder.lowerLimit.get()[y] <= upperLimit[y], "incorrect limits");
+            }
+            lowerLimit = Arrays.copyOf(builder.lowerLimit.get(), inputLength);
+        } else {
+            lowerLimit = new float[inputLength];
+            Arrays.fill(lowerLimit, -Float.MAX_VALUE);
+        }
+        // uses lastInputs as a markov input, the +1 corresponds to lookahead and then
+        // 2*inputlength correspond to the (correlated) positive and negative errors
+        //
+        if (builder.useRCF) {
+            int inputDimensions = lastInputs.length + 2 * inputLength + 2;
+            double[] weights = new double[inputDimensions];
+            Arrays.fill(weights, 1.0);
+            // the lookahead has 1/3 the weight of the total
+            weights[lastInputs.length] = lastInputs.length;
+            weights[lastInputs.length + 1] = lastInputs.length;
+            estimator = new PredictiveRandomCutForest.Builder<>().inputDimensions(inputDimensions).weights(weights)
+                    .randomSeed(13).outputAfter(50).transformMethod(TransformMethod.NORMALIZE).startNormalization(49)
+                    .build();
+        }
+    }
+
+    // for mappers
+    public ErrorHandler(int errorHorizon, int forecastHorizon, int sequenceIndex, double percentile, int inputLength,
+            float[] pastForecastsFlattened, float[] lastDataDeviations, double[] lastInput, Deviation[] deviations,
+            PredictiveRandomCutForest estimator, float[] auxiliary) {
+        checkArgument(forecastHorizon > 0, " incorrect forecast horizon");
+        checkArgument(errorHorizon >= forecastHorizon, "incorrect error horizon");
+        checkArgument(inputLength > 0, "incorrect parameters");
+        checkArgument(sequenceIndex >= 0, "cannot be negative");
+        checkArgument(Math.abs(percentile - 0.25) < 0.24, "has to be between (0,0.5) ");
+        checkArgument(deviations.length == 3 * inputLength * forecastHorizon, "incorrect length");
+        checkArgument(lastInput.length == 2 * inputLength, "incorrect length");
```

Comment:
Why do you keep last 2 input instead of 1? I know more recorded value in the estimator/model are better.

Reply:
That seemed to help the error. Suppose we were using 1 variable -- then the error has two components high and low. But the current variable and lookahead are only 2 pieces of information -- so we were imputing 2 out of 4. With twice the input length, this makes imputing 2 out of 5 -- so the noise was lessened. Did not seem to have any benefit from going to thrice ...